Update internationalization process with complete updated example. (#13909)
WordPress/gutenberg: docs/designers-developers/developers/internationalization.md

When you set script translations for a handle WordPress will automatically figure out if a translations file exists on translate.wordpress.org, and if so ensure that it's loaded into `wp.i18n` before your script runs.  With translate.wordpress.org, plugin authors also do not need to worry about setting up their own infrastructure for translations and can rely on a global community with dozens of active locales. Read more about [WordPress Translations](https://make.wordpress.org/meta/handbook/documentation/translations/).

## Provide Your Own Translations

You can create and ship your own translations with your plugin, if you have sufficient knowledge of the language(s) you can ensure the translations are available.

### Create Translation File

The translation files must be in the JED 1.x JSON format.

To create a JED translation file, first you need to extract the strings from the text. Typically, the language files all live in a directory called `languages` in your plugin.  Using [WP-CLI](https://wp-cli.org/), you create a `.pot` file using the following command from within your plugin directory:

```
mkdir languages

> wp i18n make-pot ./ languages/myguten.pot

Reviewer: I'm running into https://github.com/wp-cli/i18n-command/issues/138 when executing this command. To continue testing, what I did was to copy the below `myguten.pot` as if the command succeeded.

Author: Do you have `php_mbstring` installed?

Reviewer: That was it! After installing the PHP `mbstring` module, this worked as expected.

Although this should be noted somewhere in i18n as a dependency, would you think it'd merit a note here as well?

Reviewer 2: As noted on that issue you referenced, we might remove the mbstring dependency in the i18n command anyway.

```

This will create the file `myguten.pot` which contains all the translatable strings from your project. A single entry in this POT file might look like this:

```

> #: block.js:6

Reviewer: I think this file may be missing the other two strings? It'll be great to provide a complete file to prevent the user from being stuck if the `wp i18n` command fails, as it was in my case.

Author: Added full file examples.

msgid "Simple Block"
msgstr ""
```

Here, `msgid` is the string to be translated, and `msgstr` is the actual translation. In the POT file, `msgstr` will always be empty.

This POT file can then be used as the template for new translations. You should copy the file using the language code you are going to translate, this example will use the Esperanto (eo) language:

```
cp myguten.pot myguten-eo.po
```

For this simple example, you can simply edit the `.po` file in your editor and add the translation to all the `msgstr` sets. For a larger, more complex set of translation, the [Glotpress](https://glotpress.blog/) and [poedit](https://poedit.net/) tools exist to help.

```

> #: block.js:6

Reviewer: Same comment here about providing the complete file (I'm assuming the `Hello World` string should also be in those files).

Author: Added full file examples

msgid "Simple Block"
msgstr "Simpla bloko"
```

The last step to create the translation file is to convert the `myguten-eo.po` to the JSON format needed. For this, you use the [po2json utility](https://github.com/mikeedwards/po2json) which you can install using npm. It might be easiest to install globally using: `npm install -g po2json`. Once installed, use the following command to convert to JED format:

```

> po2json myguten-eo.po myguten-eo.json -f jed

Reviewer: I'm running into this error when executing this command:

![po2json-plural-form-errors](https://user-images.githubusercontent.com/583546/53976985-6154c500-4108-11e9-960b-a94dd2f3dc16.png)

Reviewer: We could also include the output of this command, so the user knows what to expect (and isn't stuck if the command fails for some reason).

Author: Added the JSON file which should be created.

# Internationalization

## What is Internationalization?

Internationalization is the process to provide multiple language support to software, in this case WordPress. Internationalization is often abbreviated as **i18n**, where 18 stands for the number of letters between the first _i_ and the last _n_.

> Providing i18n support to your plugin and theme allows it to reach the largest possible audience, even without requiring you to provide the additional language translations.  When you upload your software to wordpress.org, all JS and PHP files will automatically be parsed. Any detected translation strings are added to translate.wordpress.org to allow the community to translate, ensuring WordPress plugins and themes are available in as many languages as possible.

Reviewer 2: ```suggestion
Providing i18n support to your plugin and theme allows it to reach the largest possible audience, even without requiring you to provide the additional language translations.  When you upload your software to WordPress.org, all JS and PHP files will automatically be parsed. Any detected translation strings are added to [translate.wordpress.org](https://translate.wordpress.org/) to allow the community to translate, ensuring WordPress plugins and themes are available in as many languages as possible.
```

	edit: () => {
		return el(
			'p',
			{ style: { color:'red'}, },
			__('Hello World', 'myguten')
		);
	},

	save: () => {
		return el(
			'p',
			{ style: { color:'red'}, },
			__('Hello World', 'myguten')
		);
	}
});
```

In the above example, the function will use the first argument for the string to be translated. The second argument is the text domain which must match the text domain slug specified by your plugin.

Common functions available, these mirror their PHP counterparts are:

- `__( 'Hello World', 'my-text-domain' )` - Translate a certain string.
- `_n( '%s Comment', '%s Comments', numberOfComments, 'my-text-domain' )` - Translate and retrieve the singular or plural form based on the supplied number.
- `_x( 'Default', 'block style', 'my-text-domain' )` - Translate a certain string with some additional context.

**Note:** Every string displayed to the user should be wrapped in an i18n function.

After all strings in your code is wrapped, the final step is to tell WordPress your JavaScript contains translations, using the [wp_set_script_translations()](https://developer.wordpress.org/reference/functions/wp_set_script_translations/) function.

```php
<?php
	...

> 	// Added in init action

Reviewer 2: I would actually write a function here and hook it into `init`, instead of adding this just as a comment.

Author: Thanks, better solution. 👍

	// Uses script handle defined in register
	wp_set_script_translations( 'myguten-script', 'myguten' );
```

This is all you need to make your plugin JavaScript code translatable.

> When you set script translations for a handle WordPress will automatically figure out if a translations file exists on translate.wordpress.org, and if so ensure that it's loaded into wp.i18n before your script runs.  With translate.wordpress.org, plugin authors also do not need to worry about setting up their own infrastructure for translations and can rely on a global community with dozens of active locales.  Read more about [WordPress Translations](https://make.wordpress.org/meta/handbook/documentation/translations/).

Reviewer 2: ```suggestion
When you set script translations for a handle WordPress will automatically figure out if a translations file exists on translate.wordpress.org, and if so ensure that it's loaded into `wp.i18n` before your script runs.  With translate.wordpress.org, plugin authors also do not need to worry about setting up their own infrastructure for translations and can rely on a global community with dozens of active locales. Read more about [WordPress Translations](https://make.wordpress.org/meta/handbook/documentation/translations/).
```

## Provide Your Own Translations

You can create and ship your own translations with your plugin, if you have sufficient knowledge of the language(s) you can ensure the translations are available.

### Create Translation File

The translation files must be in the JED 1.x JSON format.

To create a JED translation file, first you need to extract the strings from the text.

> Using the [wp-cli tool](https://wp-cli.org/), you create a `.pot` file using the following command from within your plugin directory:

Reviewer 2: ```suggestion
Using [WP-CLI](https://wp-cli.org/), you create a `.pot` file using the following command from within your plugin directory:
```

```
wp i18n make-pot ./
```

> This will create the file `myguten.pot` a pot file is a template file to be used for translating to different languages. The pot file is made up of translation sets, for example:

Reviewer 2: ```suggestion
This will create the file `myguten.pot` which contains all the translatable strings from your project. A single entry in this POT file might look like this:
```

```
#: block.js:6
msgid "Simple Block"
msgstr ""
```

> The `msgid` is the string to be translated, and `msgstr` is the translated string, which in the pot file the msgstr will be empty.

Reviewer 2: ```suggestion
Here, `msgid` is the string to be translated, and `msgstr` is the actual translation. In the POT file, `msgstr` will always be empty.
```

> A po file is then created from this template, which is the same format, but the translations filled in. You should copy the file using the language code you are going to translate, this example will use the Esperanto (eo) language:

Reviewer 2: ```suggestion
This POT file can then be used as the template for new translations. You should copy the file using the language code you are going to translate, this example will use the Esperanto (eo) language:
```

```
cp myguten.pot myguten-eo.po
```

> You then go through the `.po` file and add the translation to all the `msgstr` sets:

Reviewer 2: I am a bit skeptical about this suggestion. For beginners it isn't the easiest file format, especially when there are more complex strings with plural forms involved.

I would at least make a note about existing software like GlotPress or Poedit.

Author: Added 👍

```
#: block.js:6
msgid "Simple Block"
msgstr "Simpla bloko"
```

> The last step to create the translation file is to convert the `myguten-eo.po` to the JSON format needed. For this, you use the [po2json utility](https://github.com/mikeedwards/po2json) which you can install using npm. It might be easiest to install globally using: `npm install -g po2json`. Once installed, use the following command to convert to JED format:

Reviewer 2: Here I would totally recommend using WP-CLI for this instead.

It's used in the previous step already and supports JSON extraction as well (`wp i18n make-json`).

Recommending po2json means yet another tool one has to install and worry about.

Author: I definitely would perfer to use the WP-CLI version, but I struggled with i and have not been able to get it to work. I've read the help, tried numerous different ways.

I wondered why the documentation here also referred to po2json: https://make.wordpress.org/core/2018/11/09/new-javascript-i18n-support-in-wordpress/

Author: Ticket opened here #14093

```
po2json myguten-eo.po myguten-eo.json -f jed
```

### Load Translation File

The final part is to tell WordPress where it can look to find the translation file. The `wp_set_script_translations` function accepts an optional third argument that is the path it will first check for translations. For example:

```php
<?php

> 	// added in init action

Reviewer 2: I would actually write a function here and hook it into `init`, instead of adding this just as a comment.

	wp_set_script_translations( 'myguten-script', 'myguten', plugin_dir_path( __FILE__ ) . 'languages' );
```

> WordPress will check for a file in that path with the format `${domain}-${locale}-${handle}.json` as the source of translations.

Reviewer 2: Might be worth mentioning the `${domain}-${locale}-${md5hash}.json` format as well, even if it's just by adding a link to the dev-note.

Author: Added 👍

> So you should create a directory called `languages` and move the `myguten-eo.json` file there, the handle in the file name is the same handle for the script that needs translating, `myguten-script` in this example:

Reviewer 2: This is also the place where the POT and PO files should live, so ideally this should be mentioned a bit further up. At least in the examples.

Author: I updated and move the directory part up. 👍

```
mkdir languages
mv myguten-eo.json myguten-eo-myguten-script.json
```

> You will need to set your WordPress installation to Esperanto language, you can do so in Settings > General and change your site language to Esperanto. This may require you add `define( 'WPLANG', 'eo' );` to your wp-config.php.

Reviewer 2: This constant is **deprecated**. One should use the locale option in Settings -> General in WordPress instead.

Author: Removed 👍
